internal/datanode/channel_manager.go

```diff
@@ -0,0 +1,505 @@
+	opsInQueue chan *datapb.ChannelWatchInfo
+	resultCh   chan *opState
+
+	closeWg   sync.WaitGroup
+	closeOnce sync.Once
+	closeCh   chan struct{}
+}
+
+func NewOpRunner(channel string, dn *DataNode, f releaseFunc, resultCh chan *opState) *opRunner {
+	return &opRunner{
+		channel:     channel,
+		dn:          dn,
+		releaseFunc: f,
+		opsInQueue:  make(chan *datapb.ChannelWatchInfo, 10),
+		allOps:      make(map[UniqueID]*opInfo),
+		resultCh:    resultCh,
+		closeCh:     make(chan struct{}),
+	}
+}
+
+func (r *opRunner) Start() {
+	r.closeWg.Add(1)
+	go func() {
+		defer r.closeWg.Done()
+		for {
+			select {
+			case info := <-r.opsInQueue:
+				r.NotifyState(r.Execute(info))
+			case <-r.closeCh:
+				return
+			}
+		}
+	}()
+}
+
+func (r *opRunner) FinishOp(opID UniqueID) {
+	r.guard.Lock()
+	defer r.guard.Unlock()
+	delete(r.allOps, opID)
+}
+
+func (r *opRunner) Exist(opID UniqueID) bool {
+	r.guard.RLock()
+	defer r.guard.RUnlock()
+	_, ok := r.allOps[opID]
+	return ok
+}
+
+func (r *opRunner) Enqueue(info *datapb.ChannelWatchInfo) error {
+	if info.GetState() != datapb.ChannelWatchState_ToWatch &&
+		info.GetState() != datapb.ChannelWatchState_ToRelease {
+		return errors.New("Invalid channel watch state")
+	}
+
+	r.guard.Lock()
+	defer r.guard.Unlock()
+	if _, ok := r.allOps[info.GetOpID()]; !ok {
+		r.opsInQueue <- info
+		r.allOps[info.GetOpID()] = &opInfo{}
+	}
+	return nil
+}
+
+func (r *opRunner) UnfinishedOpSize() int {
+	r.guard.RLock()
+	defer r.guard.RUnlock()
+	return len(r.allOps)
+}
+
+// Execute excutes channel operations, channel state is validated during enqueue
+func (r *opRunner) Execute(info *datapb.ChannelWatchInfo) *opState {
+	log.Info("Start to execute channel operation",
+		zap.String("channel", info.GetVchan().GetChannelName()),
+		zap.Int64("opID", info.GetOpID()),
+		zap.String("state", info.GetState().String()),
+	)
+	if info.GetState() == datapb.ChannelWatchState_ToWatch {
+		return r.watchWithTimer(info)
+	}
+
+	// ToRelease state
+	return releaseWithTimer(r.releaseFunc, info.GetVchan().GetChannelName(), info.GetOpID())
+}
+
+// watchWithTimer will return WatchFailure after WatchTimeoutInterval
+func (r *opRunner) watchWithTimer(info *datapb.ChannelWatchInfo) *opState {
+	opState := &opState{
+		channel: info.GetVchan().GetChannelName(),
+		opID:    info.GetOpID(),
+	}
+	log := log.With(zap.String("channel", opState.channel), zap.Int64("opID", opState.opID))
+
+	r.guard.Lock()
+	opInfo, ok := r.allOps[info.GetOpID()]
+	if !ok {
+		opState.state = datapb.ChannelWatchState_WatchFailure
+		return opState
+	}
+	tickler := newTickler()
+	opInfo.tickler = tickler
+	r.guard.Unlock()
+
+	var (
+		successSig = make(chan struct{}, 1)
+		waiter     sync.WaitGroup
+	)
+
+	watchTimeout := Params.DataCoordCfg.WatchTimeoutInterval.GetAsDuration(time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), watchTimeout)
+	defer cancel()
+
+	startTimer := func(wg *sync.WaitGroup) {
```

Comment:
timer should be a single instance?
We don't need to create one goroutine for each timer.

There should be a global ticker and process tickler for each job